src/Bicep.Core/TypeSystem/TypeValidationVisitor.cs

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.Syntax;
+using Bicep.Core.Syntax.Visitors;
+
+namespace Bicep.Core.TypeSystem
+{
+    public sealed class TypeValidationVisitor : SyntaxVisitor
+    {
+        private readonly IReadOnlyDictionary<SyntaxBase, Symbol> bindings;
+        private readonly ITypeManager typeManager;
+        private readonly IList<Diagnostic> diagnostics;
+        private readonly IDictionary<SyntaxBase, ImmutableArray<Diagnostic>> syntaxDiagnostics;
```

Comment:
>syntaxDiagnostics [](start = 77, length = 17)

Does this need to be a dictionary? It seems we're not actually looking up anything by key after putting in the diagnostics per syntax node.

Reply:
Good point. I thought I needed it, but turned out I didn't. Will remove this in next iteration

Reply:
Sounds good.

---
In reply to: [493013888](https://github.com/Azure/bicep/pull/546#discussion_r493013888) [](ancestors = 493013888)